src/Connection/DefaultConnectionPool.php

```diff
@@ -51,27 +55,47 @@ public function getStream(Request $request, CancellationToken $cancellation): Pr
             $port = $uri->getPort() ?? $defaultPort;
 
             if ($host === '') {
-                throw new HttpException('A host must be provided in the request URI: ' . $uri);
+                throw new InvalidRequestException($request, 'A host must be provided in the request URI: ' . $uri);
             }
 
             $authority = $host . ':' . $port;
             $key = $scheme . '://' . $authority;
 
             if (!\array_intersect($request->getProtocolVersions(), self::PROTOCOL_VERSIONS)) {
-                throw new HttpException('None of the requested protocol versions are supported; Supported versions: '
-                    . \implode(', ', self::PROTOCOL_VERSIONS));
+                throw new InvalidRequestException(
+                    $request,
+                    'None of the requested protocol versions are supported; Supported versions: '
+                        . \implode(', ', self::PROTOCOL_VERSIONS)
+                );
+            }
+
+            if (!$isHttps && !\array_intersect($request->getProtocolVersions(), ['1.0', '1.1'])) {
+                throw new InvalidRequestException(
+                    $request,
+                    'HTTP/1.x forbidden, but a secure connection (HTTPS) is required for HTTP/2'
+                );
             }
 
-            $connections = $this->connections[$key] ?? new \SplObjectStorage;
+            $i = 0;
+            while ($hashes = \array_keys($this->connections[$key] ?? [])) {
+                if (isset($hash) && $hash !== $hashes[$i++]) {
+                    $i = 0; // Connection list was modified while waiting, start over from beginning.
+                }
+
+                if (!isset($hashes[$i])) {
+                    break;
+                }
+
+                $hash = $hashes[$i];
```

Comment:
This is very hard to read now.

Reply:
The only thing I can immediately think to do is move the increment on `$i`. Suggestions?

Reply:
Do we really care about modifications while iterating here? The only case where we really wait is having one connection and that one is pending.

Reply:
True, now that we only wait with one connection there isn't a chance for modifications while waiting. That will make this simple.